extension/source/cells/CellCollection.ts

```diff
@@ -0,0 +1,206 @@
+import { EventEmitter } from 'events';
+
+import * as io from 'io-ts';
+
+import ExplicitAny from '../types/ExplicitAny';
+import ICell, { CellEmitter } from './ICell';
+import CellIterator from './CellIterator';
+import jsonHasChanged from './jsonHasChanged';
+import assert from '../helpers/assert';
+import IAsyncStorage from './IAsyncStorage';
+
+export default class CellCollection {
+  cells: Record<string, CollectionCell<ExplicitAny> | undefined> = {};
+
+  #asyncStorageChangeHandler = (keys: string[]) => {
+    for (const key of keys) {
+      const cell = this.cells[key];
+
+      if (cell) {
+        cell.versionedRead();
+      }
+    }
+  };
+
+  constructor(public asyncStorage: IAsyncStorage) {
+    asyncStorage.events.on('change', this.#asyncStorageChangeHandler);
+  }
+
+  /**
+   * A `CellCollection` is usually intended to live for the life of your
+   * application. In that case, calling .end is not required. However, if you're
+   * doing something different and you'd like to clean up the change event
+   * handler, this will take care of that.
+   */
+  end() {
+    this.asyncStorage.events.removeListener(
+      'change',
+      this.#asyncStorageChangeHandler,
+    );
+  }
+
+  Cell<T>(
+    key: string,
+    type: io.Type<T>,
+    defaultValue: T,
+    hasChanged = jsonHasChanged,
+  ): CollectionCell<T> {
+    let cell = this.cells[key];
+
+    if (cell) {
+      cell.applyType(type);
+      return cell;
+    }
+
+    cell = new CollectionCell(
+      this.asyncStorage,
+      key,
+      type,
+      defaultValue,
+      hasChanged,
+    );
+
+    this.cells[key] = cell;
+
+    return cell;
+  }
+
+  async remove(key: string) {
+    const cell = this.cells[key];
+    cell?.end();
+    delete this.cells[key];
+
+    await this.asyncStorage.write(key, io.undefined, undefined);
+  }
+}
+
+export type Versioned<T> = { version: number; value: T };
+
+export class CollectionCell<T> implements ICell<T> {
+  events = new EventEmitter() as CellEmitter<T>;
+  ended = false;
+
+  versionedType: io.Type<Versioned<T>>;
+  lastSeen?: Versioned<T>;
+  initialRead: Promise<void>;
+
+  constructor(
+    public asyncStorage: IAsyncStorage,
+    public key: string,
+    public type: io.Type<T>,
+    public defaultValue: T,
+    public hasChanged: ICell<T>['hasChanged'] = jsonHasChanged,
+  ) {
+    this.versionedType = io.type({ version: io.number, value: type });
+
+    this.initialRead = this.versionedRead().then((versionedReadResult) => {
+      this.lastSeen = versionedReadResult;
+    });
```

Comment:
I think this is ok. The checks in `applyType` are just proactive checks to help raise errors earlier/closer to where the mistake is made. Your app will still break either way if the types are not consistent.

If `initialRead` completes before `applyType` then you'll get the benefit of `applyType` throwing an exception if the concrete value doesn't match that type.

If `applyType` is run before `initialRead` completes then `initialRead` will be rejected if the concrete value doesn't match the type (which will show in tools/console) and all writes will also fail with this error since they have `await this.initialRead;`.